# Tidy Data

In [1]:
import pandas as pd
import numpy as np

### Problem 1

Attendance Data

Load the attendance.csv file and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.

In [2]:
df_attendance = pd.read_csv('attendance.csv')

In [3]:
df_attendance

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [4]:
def get_attendance_grade(letter):
    if letter == 'P':
        return 1
    elif letter == 'A':
        return 0
    elif letter == 'T':
        return 0.9
    elif letter == 'H':
        return 0.5

get_attendance_grade('T')

0.9

In [5]:
grades = {
    'P': 1,
    'A': 0,
    'H': 0.5,
    'T': 0.9,
}

df = pd.read_csv('attendance.csv')
df = df.rename(columns={'Unnamed: 0': 'student'})
df = df.melt(id_vars='student', var_name='date')
df['grade'] = df.value.apply(get_attendance_grade)
df.groupby('student').grade.mean()

student
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64

### Problem 2

Coffee Levels

a. Read the coffee_levels.csv file.

In [6]:
df_coffee = pd.read_csv('untidy-data/coffee_levels.csv')

In [7]:
df_coffee.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

b. Transform the data so that each carafe is in it's own column.

In [8]:
df_coffee = df_coffee.pivot_table(values='coffee_amount', index='hour', columns='coffee_carafe')
df_coffee.round(3)

coffee_carafe      x      y      z
hour                              
8              0.816  0.189  0.999
9              0.451  0.522  0.916
10             0.843  0.023  0.145
11             0.336  0.236  0.311
12             0.898  0.017  0.772
13             0.311  0.997  0.399
14             0.507  0.058  0.864
15             0.215  0.145  0.436
16             0.184  0.545  0.281
17             0.392  0.594  0.437

c. Is this the best shape for the data?

Yes, it is much easier to read and understand now

### Problem 3

Cake Recipes 

a. Read the cake_recipes.csv data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.

In [9]:
df_cake = pd.read_csv('untidy-data/cake_recipes.csv')
df_cake.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

b. Tidy the data as necessary.

In [10]:
recipes_and_positions = df_cake['recipe:position'].str.split(':', expand=True)
recipes_and_positions.columns = ['recipe', 'position']
df_cake = pd.concat([df_cake, recipes_and_positions], axis=1).drop(columns='recipe:position')
df_cake = df_cake.melt(id_vars=['recipe', 'position'], var_name='temp', value_name='tastiness')
df_cake.head()

recipe position temp  tastiness
0      a   bottom  225  61.738655
1      a      top  225  51.709751
2      b   bottom  225  57.095320
3      b      top  225  82.455004
4      c   bottom  225  96.470207

c. Which recipe, on average, is the best? recipe b 

In [11]:
df_cake.groupby('recipe').tastiness.mean()

recipe
a    63.922201
b    76.736074
c    75.874748
d    62.864844
Name: tastiness, dtype: float64

d. Which oven temperature, on average, produces the best results? 275 

In [12]:
df_cake.groupby('temp').tastiness.mean()

temp
225    71.306022
250    66.577437
275    74.886754
300    66.627655
Name: tastiness, dtype: float64

e. Which combination of recipe, rack position, and temperature gives the best result? recipe b, bottom rack, 300 degrees  

In [13]:
df_cake.sort_values(by='tastiness').tail()

recipe position temp  tastiness
4       c   bottom  225  96.470207
7       d      top  225  96.873178
19      b      top  275  98.594881
24      a   bottom  300  98.786784
26      b   bottom  300  99.248541